AY2021S1-CS2103-T16-3/tp
# Improve room filtering

### What this does
This PR allows Rooms to be filtered by floor, room number, room type and vacancy using the existing `rooms` command. Multiple filters can be specified. For example, `rooms floor/11 floor/12` will show all rooms on level 11 and 12, `rooms floor/11 floor/12 type/CA type/CN` will show all corridor rooms on levels 11 and 12.  

In addition, any commands which modify the students, rooms or allocations, including undo and redo, will now maintain the filtering instead of resetting the filters as before. This includes commands and filters pertaining to students, although this isn't visible since there are currently no commands for filtering students.

Tests for the new filtering subcommands are included.

To be done in another PR:
- Update `rooms` command in user guide
- Update use cases. 

Closes #38, closes #39.

### How to test
1.  Delete `addressbook.json` and start the app.
1.  Run the command `rooms floor/11 floor/9`. Only rooms 11-108, 11-104 and 9-102 should be shown.
1.  Run the command `rooms floor/11 floor/7 type/CN`. Only rooms 11-104 and 7-106 should be shown.
To verify that the filters do not get reset and are updated automatically:
1. Run `rooms --vacant`. Only the 4 unallocated rooms should be shown.
1. Run `allocate ri/1 si/2`. There should only be 3 unallocated rooms shown now.
1. Run `undo`. Verify the rooms list is the same as in step 4.
1. Run `redo`.  Verify the rooms list is the same as in step 5.

### Notes
- I extracted out the `arePrefixesPresent` method used in several command parsers into `ParserUtil`, so there's no need to copy the function across parsers.
